refactored argmin and argmax into iterator extension traits and moved…

zommiommy opened the pull request:
I cleaned up the implementations of arg{min, max} and arg{min, max}_filtered.

Argmin and argmax are helpful in general, so I refactored them into an extension trait for Iterator.

Meanwhile, arg_{min, max}_filtered are only used by computer.rs, so I moved them there.

zommiommy committed:
refactored argmin and argmax into iterator extension traits and moved argmin_filtered and argmax_filtered inside computer.rs as it's the only user

vigna commented:
There are literally hundreds of formatting changes... weird, I run `cargo fmt` all the time. How could this happen?

vigna commented:
> I cleaned up the implementations of arg{min, max} and arg{min, max}_filtered.
> 
> Argmin and argmax are helpful in general, so I refactored them into an extension trait for Iterator.
> 
> Meanwhile, arg_{min, max}_filtered are only used by computer.rs, so I moved them there.

Well, also `arg_min` and `arg_max` are used only by computer.rs...

zommiommy commented:
> There are literally hundreds of formatting changes... weird, I run `cargo fmt` all the time. How could this happen?

Maybe different cargo versions (nightly ?). I can re-make the PR to solve this.

> > I cleaned up the implementations of arg{min, max} and arg{min, max}_filtered.
> > Argmin and argmax are helpful in general, so I refactored them into an extension trait for Iterator.
> > Meanwhile, arg_{min, max}_filtered are only used by computer.rs, so I moved them there.
> 
> Well, also `arg_min` and `arg_max` are used only by computer.rs...

Sure, but they are more general, the thing I care the most in this PR is the simplified implementations like:
```rust
pub fn argmin_filtered<
    T: std::cmp::PartialOrd + Copy,
    N: std::cmp::PartialOrd + Copy,
    F: Fn(usize, T) -> bool,
>(
    iter: impl IntoIterator<Item = T>,
    tie_break: impl IntoIterator<Item = N>,
    filter: F,
) -> Option<usize> {
    iter.into_iter()
        .zip(tie_break)
        .enumerate()
        .filter(|v| filter(v.0, v.1 .0))
        .min_by(|a, b| {
            let (value_a, tie_a) = a.1;
            let (value_b, tie_b) = b.1;
            value_a
                .partial_cmp(&value_b)
                .unwrap()
                .then(tie_a.partial_cmp(&tie_b).unwrap())
        })
        .map(|m| m.0)
}
```
becomes:
```rust
fn argmin_filtered<I: IntoIterator, T: IntoIterator>(
    iter: I,
    tie_break: T,
    filter: impl Fn(usize, I::Item) -> bool,
) -> Option<usize>
where
    I::Item: std::cmp::PartialOrd + Copy,
    T::Item: std::cmp::PartialOrd + Copy,
{
    iter.into_iter()
        .zip(tie_break)
        .enumerate()
        .filter(|(idx, (v, _tie))| filter(*idx, *v))
        .min_by(|(_, a), (_, b)| b.partial_cmp(a).unwrap())
        .map(|(idx, _)| idx)
}
```

vigna commented:
It's much better of course, but we need to eliminate the massive amount of spacing edits first.

zommiommy commented:
Ok when I'll land I'll open a new PR without the format problem
